packages/core/src/application.ts

```diff
@@ -446,8 +483,51 @@ export class Application extends Context implements LifeCycleObserver {
         }
       }
     };
-    // eslint-disable-next-line @typescript-eslint/no-misused-promises
-    signals.forEach(sig => process.on(sig, cleanup));
+    this.registerSignalListener();
+    return this._signalListener;
+  }
+
+  private registerSignalListener() {
+    const {signals = []} = this._shutdownOptions;
+    debugShutdown(
+      '[%s] Registering signal listeners on the process %d',
+      this.name,
+      process.pid,
+      signals,
+    );
+    signals.forEach(sig => {
+      if (process.getMaxListeners() <= process.listenerCount(sig)) {
+        if (debugWarning.enabled) {
+          debugWarning(
+            '[%s] %d %s listeners are added to process %d',
+            this.name,
+            process.listenerCount(sig),
+            sig,
+            process.pid,
+            new Error('MaxListenersExceededWarning'),
+          );
+        }
+      }
+      // eslint-disable-next-line @typescript-eslint/no-misused-promises
+      process.removeListener(sig, this._signalListener);
```

Comment:
I am confused, why are we removing the listener and then adding it again on the next line? Isn't it enough to addd it?

Reply:
The initial intention was to make sure there is no duplicate registration of the same listener. I removed it now.